src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,310 @@ public int GetCount(bool onlyIfCheap)
                 return Count;
             }
         }
+
+        private sealed class EnumerablePartition<TSource> : Iterator<TSource>, IPartition<TSource>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _minIndexInclusive;
+            private readonly int _maxIndexInclusive; // -1 if we want everything past _minIndexInclusive.
+                                                     // If this is -1, it's impossible to set a limit on the count.
+            private IEnumerator<TSource> _enumerator;
+
+            internal EnumerablePartition(IEnumerable<TSource> source, int minIndexInclusive, int maxIndexInclusive)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(!(source is IList<TSource>), $"The caller needs to check for {nameof(IList<TSource>)}.");
+                Debug.Assert(minIndexInclusive >= 0);
+                Debug.Assert(maxIndexInclusive >= -1);
+                // Note that although maxIndexInclusive can't grow, it can still be int.MaxValue.
+                // We support partitioning enumerables with > 2B elements. For example, e.Skip(1).Take(int.MaxValue) should work.
+                // But if it is int.MaxValue, then minIndexInclusive must != 0. Otherwise, our count may overflow.
+                Debug.Assert(maxIndexInclusive == -1 || (maxIndexInclusive - minIndexInclusive < int.MaxValue), $"{nameof(Limit)} will overflow!");
+                Debug.Assert(maxIndexInclusive == -1 || minIndexInclusive <= maxIndexInclusive);
+
+                _source = source;
+                _minIndexInclusive = minIndexInclusive;
+                _maxIndexInclusive = maxIndexInclusive;
+            }
+
+            // If this is true (e.g. at least one Take call was made), then we have an upper bound
+            // on how many elements we can have.
+            private bool HasLimit => _maxIndexInclusive != -1;
+
+            private int Limit => (_maxIndexInclusive + 1) - _minIndexInclusive; // This is that upper bound.
+
+            public override Iterator<TSource> Clone()
+            {
+                return new EnumerablePartition<TSource>(_source, _minIndexInclusive, _maxIndexInclusive);
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
```

Comment:
@stephentoub Made the change. Something to note: While adding coverage for this, I also realized I had forgotten to override `Select` in this type. ([`Select`](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Select.cs#L25) checks for an Iterator first, then an IPartition, and the default implementation of `Iterator.Select` is to return a `SelectEnumerableIterator` rather than a `SelectIPartitionIterator`.) So thank you for pointing that out.